helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java

```diff
@@ -349,4 +368,63 @@ private IdealState generateIdealStateWithAssignment(String resourceName,
+   * @param resourceMap              the rebalanaced resource map.
+   * @param clusterChanges           the detected cluster changes that triggeres the rebalance.
+   * @param resourceStatePriorityMap the state priority map for each resource.
+   */
+  private void delayedRebalance(Map<String, IdealState> idealStateMap,
+      ResourceControllerDataProvider clusterData, Set<String> activeInstances,
+      Map<String, Resource> resourceMap, Map<HelixConstants.ChangeType, Set<String>> clusterChanges,
+      Map<String, Map<String, Integer>> resourceStatePriorityMap) throws HelixRebalanceException {
+    ClusterConfig clusterConfig = clusterData.getClusterConfig();
+    long delay = clusterConfig.getRebalanceDelayTime();
+    Set<String> delayedActiveNodes = DelayedRebalanceUtil
+        .getActiveInstances(clusterData.getAllInstances(), activeInstances,
+            clusterData.getInstanceOfflineTimeMap(), clusterData.getLiveInstances().keySet(),
+            clusterData.getInstanceConfigMap(), delay, clusterConfig);
+
+    if (!delayedActiveNodes.equals(activeInstances)) {
+      // Calculate new assignment based on delayed active node set.
+      // This additional assignment will be used to adjust the finalIdealState according to the delayed active nodes.
+      // Note that the calculation used the baseline as the input only. This is for minimizing unnecessary partition movement.
+      Map<String, ResourceAssignment> activeAssignment =
+          calculateAssignment(clusterData, clusterChanges, resourceMap, delayedActiveNodes,
+              Collections.emptyMap(), _assignmentMetadataStore.getBaseline());
+      for (String resourceName : idealStateMap.keySet()) {
+        IdealState is = idealStateMap.get(resourceName);
+        if (activeAssignment.containsKey(resourceName)) {
+          IdealState currentIdealState = clusterData.getIdealState(resourceName);
+          IdealState newActiveIdealState =
+              generateIdealStateWithAssignment(resourceName, currentIdealState,
+                  activeAssignment.get(resourceName), resourceStatePriorityMap.get(resourceName));
+          Map<String, List<String>> finalPreferenceList = DelayedRebalanceUtil
+              .getFinalDelayedMapping(is.getPreferenceLists(),
+                  newActiveIdealState.getPreferenceLists(), activeInstances, DelayedRebalanceUtil
+                      .getMinActiveReplica(currentIdealState,
+                          currentIdealState.getReplicaCount(delayedActiveNodes.size())));
+          is.setPreferenceLists(finalPreferenceList);
+        }
+      }
+    }
+  }
+
+  private void applyUserDefinePreferenceList(ResourceConfig resourceConfig, IdealState idealState) {
+    if (resourceConfig != null) {
+      Map<String, List<String>> userDefinedPreferenceList = resourceConfig.getPreferenceLists();
+      if (!userDefinedPreferenceList.isEmpty()) {
+        LOG.info("Using user defined preference list for partitions: " + userDefinedPreferenceList
```

Comment:
Wouldn't this be too verbose?

Wouldn't "Using user defined preference list for partitions" be enough for our purposes? Because we would have the preference list in ResourceConfig.

Reply:
Makes sense. Will change.